src/ripple/app/main/Application.cpp

```diff
@@ -171,10 +175,49 @@ class AppFamily : public shamap::Family
     void
     missing_node (std::uint32_t seq) override
     {
-        uint256 const hash = getApp().getLedgerMaster ().getHashBySeq (seq);
-        if (hash.isZero())
+        WriteLog (lsERROR, Ledger) << "Missing node in " << seq;
+
+        // prevent recursive invocation
+        std::unique_lock <std::mutex> lock (maxSeqLock);
```

Comment:
Without this change, the missing node handler could call getHashBySeq which could call the missing node handler, which could call getHashBySeq, boom. With this code, if getHashBySeq is called by the missing node handler, maxSeq (the thing the lock protects) will be non-zero, inhibiting the call to getHashBySeq and preventing the recursive invocation of the missing node handler.